File: old/sorting/quick_sort/implement_quick_sort_lomuto/implement_quick_sort_lomuto.py
def quick_sort_lomuto(arr, start, end):
    if start >= end:
        return

    left = start
    pivot = arr[end]

    for i in range(start, end):
        if arr[i] < pivot:
            temp = arr[left]
            arr[left] = arr[i]
            arr[i] = temp
            left += 1

    arr[end] = arr[left]
    arr[left] = pivot

    quick_sort_lomuto(arr, start, left - 1)
    quick_sort_lomuto(arr, left + 1, end)


def main():
    input = [1, 3, 2, 5, 4, 7, 8, 6, 9, 10]
    print(input)
    quick_sort_lomuto(input, 0, 9)
    print(input)


main()

File: old/sorting/quick_sort/implement_quick_sort_lomuto/implement_quick_sort_lomuto.cpp
#include <iostream>
#include <vector>

using namespace std;

void quickSortLomuto(vector<int>& arr, int start, int end) {
    if (end - start <= 0) {
        return;
    }
    
    int pivot = arr[end];
    int left = start;

    for (int i = start; i < end; i++) {
        if (arr[i] < pivot) {
            swap(arr[i], arr[left]);
            left++;
        }
    }

    arr[end] = arr[left];
    arr[left] = pivot;

    quickSortLomuto(arr, start, left - 1);
    quickSortLomuto(arr, left + 1, end);
}


int main() {
    vector<int> input = {0, 2, 1, 5, 4, 3, 7, 9, 8, 10, 6};
    for (int i = 0; i < input.size(); i++) {
        cout << input[i] << " ";
    }
    cout << endl;
    quickSortLomuto(input, 0, 10);
    for (int i = 0; i < input.size(); i++) {
        cout << input[i] << " ";
    }
    return 0;
}
